reana_workflow_controller/rest/utils.py

```diff
@@ -25,38 +25,66 @@
 from reana_commons.utils import get_workflow_status_change_verb
 from reana_db.database import Session
 from reana_db.models import Job, JobCache, Workflow, WorkflowStatus
+from sqlalchemy.exc import SQLAlchemyError
+from kubernetes.client.rest import ApiException
 
 from reana_workflow_controller.config import (REANA_GITLAB_HOST,
                                               WORKFLOW_TIME_FORMAT)
-from reana_workflow_controller.errors import (REANAWorkflowControllerError,
+from reana_workflow_controller.errors import (REANAExternalCallError,
+                                              REANAWorkflowControllerError,
                                               REANAWorkflowDeletionError)
 from reana_workflow_controller.workflow_run_manager import \
     KubernetesWorkflowRunManager
 
 
 def start_workflow(workflow, parameters):
     """Start a workflow."""
-    if workflow.status in [WorkflowStatus.created, WorkflowStatus.queued]:
+    current_db_sessions = Session.object_session(workflow)
+    kwrm = KubernetesWorkflowRunManager(workflow)
+
+    def _start_workflow_db(workflow, parameters):
         workflow.run_started_at = datetime.now()
         workflow.status = WorkflowStatus.running
         if parameters:
             workflow.input_parameters = parameters.get('input_parameters')
             workflow.operational_options = \
                 parameters.get('operational_options')
-        current_db_sessions = Session.object_session(workflow)
         current_db_sessions.add(workflow)
         current_db_sessions.commit()
-        kwrm = KubernetesWorkflowRunManager(workflow)
-        kwrm.start_batch_workflow_run()
-    else:
+
+    def _rollback_start_workflow_db(workflow):
```

Comment:
Would it be worth to add `retry` decorator as in https://github.com/reanahub/reana-job-controller/blob/2688996c44c2a1b275076dc58990ad259cd9db7a/reana_job_controller/htcondorcern_job_manager.py#L191? Since this is really crucial function.

Reply:
Right now there would be a retry happening the following way:
```console
RS                     RWC                    DBoD       MQ
|                       |                       |         | Q = [X]
|<----------------------|-[RS consumes wf X]----|---------| Q = []
|--[start workflow X]-->|                       |         |
|                       | --[commit to DB]--->  |         |
|                       | --* Fails to          |         | 
|                       |   | connect to DB     |         |
|                       |   | (or other failure)|         |
|                       | <-*                   |         |
|<--------[502]---------|                       |         |
|-----------------------|-[requeues workflow X]-|-------->| Q = [X]
|<----------------------|-[RS consumes wf X]----|---------| Q = []
|---[start workflow X]->|                       |
|                       |----[commit to DB]---->|
|                       |<---[success]----------|
|<--------[200]---------|
```

It is true that if we do it here, in RWC code, we avoid all this going back and forth... But this requeueing mechanism we need for any other failure, so it can serve the two purposes, WDYT? We could ticketise the `retry` for later since no other service will call RWC except RS.